lib/deimos/active_record_consume/batch_consumption.rb

```diff
@@ -154,14 +155,38 @@ def import_associations(entities)
               sub_records = Array(entity.send(assoc.name))
               # Set IDS from master to each of the records in `has_one` or `has_many` relation
               sub_records.each { |d| d.send("#{assoc.foreign_key}=", entity.send(assoc.active_record_primary_key)) }
+
+              # Retrieve bulk_import_id data to deleting repeated lines
+              delete_map = mapping_delete_query(entity, assoc, delete_map)
+
               sub_records
             }.flatten
 
             columns = key_columns(nil, assoc.klass)
             save_records_to_database(assoc.klass, columns, sub_records) if sub_records.any?
+            removing_repeated_associations(delete_map, columns, assoc.klass)
           end
       end
 
+      def mapping_delete_query(entity, assoc, delete_map)
+        return delete_map unless entity.attributes['bulk_import_id'].present?
+
+        delete_map.push('bulk_import_id' => entity.attributes['bulk_import_id'],
+                        'primary_key_value' => entity.attributes[assoc.active_record_primary_key],
+                        'primary_key_name' => assoc.foreign_key)
+        delete_map
+      end
+
+      def removing_repeated_associations(delete_map, columns, assoc)
+        return unless columns.include?('bulk_import_id') && delete_map.any?
+
+        delete_map.each do |d|
+          assoc.unscoped.
+            where(d['primary_key_name'] => d['primary_key_value']).
+            where.not('bulk_import_id' => d['bulk_import_id']).delete_all
+        end
```

Comment:
This is going to have a query _per record_ which kind of puts us back where we were. My idea here was to add a bulk\_import\_id to the _subclass_ where every sub-record would have the same bulk\_import\_id when we import them. Then after we do the import, we just need to do a single delete\_all on a per _table_ basis rather than a per _record_ basis.